# Homework

In [36]:
import os
import pickle
import click
import pandas as pd
import numpy as np

In [37]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error
from sklearn.feature_extraction import DictVectorizer

In [38]:
from hyperopt import STATUS_OK, Trials, fmin, hp, tpe
from hyperopt.pyll import scope

In [ ]:
from pathlib import Path

ROOT = Path.cwd()
TRACKING_URI = "http://127.0.0.1:5000"
EXPERIMENT_NAME = "nyc-taxi-experiment"
RAW_DATA_PATH = ROOT / "data"
OUTPUT_PATH = ROOT / "output"

Working directory: /workspaces/mlops-zoomcamp/02-experiment-tracking/Homework
Tracking URI: http://127.0.0.1:5000
Experiment: nyc-taxi-experiment
Preprocess script path: /workspaces/mlops-zoomcamp/02-experiment-tracking/Homework/preprocess_data.py
Train script path: /workspaces/mlops-zoomcamp/02-experiment-tracking/Homework/train.py


Q1. Install MLflow

In [40]:
import mlflow

mlflow.__version__

'1.25.1'

Answer: 1.25.1

Q2. Download and preprocess the data

In [41]:
from importlib import reload
import preprocess_data

reload(preprocess_data)
preprocess_data.run_data_prep.callback(
    raw_data_path=str(RAW_DATA_PATH),
    dest_path=str(OUTPUT_PATH),
    dataset="green",
)

print("Preprocess completed successfully.")

Preprocess completed successfully.


Answer: 4

Q3. Train a model with autolog

In [42]:
import train

os.environ["MLFLOW_TRACKING_URI"] = TRACKING_URI
os.environ["MLFLOW_EXPERIMENT_NAME"] = EXPERIMENT_NAME

reload(train)
train.run_train.callback(data_path=str(OUTPUT_PATH))

print("Training completed and run logged to MLflow.")

2026/07/06 12:17:35 INFO mlflow.tracking.fluent: Experiment with name 'nyc-taxi-experiment' does not exist. Creating a new experiment.


2026/07/06 12:17:35 INFO mlflow.tracking.fluent: Autologging successfully enabled for sklearn.


RMSE: 5.431162180141208
Training completed and run logged to MLflow.


Answer: 2

Q4 Launch the tracking server locally

Answer: default-artifact-root

Q5 Tune model hyperparameters

In [43]:
from importlib import reload
import hpo

reload(hpo)
hpo.run_optimization.callback(data_path=str(OUTPUT_PATH), num_trials=15)

print("Hyperparameter optimization completed and logged to MLflow.")

  0%|          | 0/15 [00:00<?, ?trial/s, best loss=?]

100%|██████████| 15/15 [01:47<00:00,  7.14s/trial, best loss: 5.335419588556921]
Parent run: 1d86b218640a4c84903a6bb59d90e35c
Best RMSE: 5.335419588556921
Best result: {'max_depth': 19.0, 'min_samples_leaf': 2.0, 'min_samples_split': 2.0, 'n_estimators': 11.0}
Hyperparameter optimization completed and logged to MLflow.


Answer: 5.335

Q6. Promote the best model to the model registry

In [44]:
from importlib import reload
import register_model

reload(register_model)
register_model.run_register_model.main(args=[], standalone_mode=False)

print("Model selection and registration completed.")

Registered model 'nyc-taxi-rf-best-model' from run f03e55e44c174a7a96793f498f91dac6
Model version: 2
Model selection and registration completed.


Answer: 5.567